### PR TITLE
[WIP] Fix 9

### PR DESCRIPTION
**Pull request by mb2055**
Fix for issue #9 . Changes the following:
- Refactor for HMR means that it is now performed in `_dynamics.py`, specifically _after_ equilibration (if applicable).
- Adds additional minimisation between equilibration and production.
- If `restart==True` and equilibration is requested, somd2 will now only equilibrate the systems that are starting from scratch, all systems restarting from file will be subject only to minimisation.
- Somd2 will now exit upon instances of file overwriting, rather than requesting user input, this can be bypassed by the new `overwrite` option (which replaces the `supress_overwrite_warning` option). Also made the logger warninig for this particular interaction more obvious.

### CODE REVIEW

**Comment by lohedges** [comment]
Thanks for this. For the HMR part could we also consider this alternative option: Keep things how they were originally, but do the following:

```python
# Store the regular system.
self._system = system

# Now do the HMR bit.
self._system_hmr = ...
```

For equilibration we use `self._system`, for production we use `self._system_hmr`. I'm not sure of the additional overhead from storing two systems, especially when using futures, but this means that we only perform HMR _once_ and any errors will be caught at the moment the `Runner` object is instantiated, rather than after equilibration, which could potentially be a long time. It might be worth experimenting with both approaches to see which one works best. In general I prefer to do things at the outset so that errors are caught early, and also avoid the overhead of duplicate operations.

Cheers.

**Comment by mb2055** [comment]
As you say, my main concern was potential slowdown from storing multiple systems in each child process. I'll do some benchmarking tomorrow on some large systems and check if there's a noticeable difference in the two approaches.

Regardless, there will still need to be HMR checks in `dynamics` for systems that are restarting from file as these won't be using the base system (systems that restart from file will skip equilibration so there's no need to store a pre-HMR version).

**Comment by chryswoods** [comment]
There should be little to no overhead in storing multiple sire systems. Internally they use copy-on-write pointers. So copying an object (anything derived from Property) is super-quick, and there is no increase in memory. The object is only really copied when you edit it. This extends down through containers like System or SelectorMol. So, copying a System will be very fast, and not increase memory. Then changing the System will create a new copy, but those copies will be shared copy-on-write copies of the contents (so super-fast, and not take up any extra memory). Only the things you actually change in System will consume more memory.

So, in the case of HMR, the "old" System that hasn't had HMR applied will use the same memory for nearly everything as the "new" System that has HMR applied, except for the AtomMasses object that holds the masses of the atoms of the molecule that has been changed.

TL;DR - never worry about storing copies of sire objects. It is designed originally as a Monte Carlo program, so copying objects is very fast, memory requirements are minimised, and only the differences between two copies actually result in an increase in memory consumption.

**Comment by mb2055** [comment]
Latest push has implemented dual-systems for HMR, but has also introduced significany complexity to the code, namely the need for special cases in both minimisation and equilibration as well as the need to robustly copy one system into another. 
Also, systems that were previously running without issue are now showing instability, suggesting that, while the tests are passing, there is an issue somewhere.
Marked as WIP due to the above.

**Comment by lohedges** [comment]
Thanks for the update. I'll hold off reviewing until you debug further. Happy to look into it with you tomorrow if needed.

**Comment by mb2055** [comment]
Following today's discussion this PR can be significantly simplified - all it now needs to do is add a brief post-equilibration minimisation (if moving to a larger timestep).

**Comment by lohedges** [comment]
Yes, I think so. I'm not sure if we want to do a two-stage equilibration, i.e. do a short run with a small time step and no constraints, then equilibrate longer using the production settings? It might be easier for you to scratch this branch and implement things with a new fork from `main`?

**Comment by mb2055** [comment]
It is definitely easier to start from a new branch.
 As for the two-stage equilibration, the question is how to implement it and make it clear with the allowed input arguments. I think the pre-existing `equilibration-time` argument should remain as it is (i.e. it sets the amount of equilibration done at a lower timestep). We could, for example, add an `ignore-time` that defines the length of the second equilibration  under the guise of ignoring the first x amount of simulation when recoding energies.

**Comment by lohedges** [comment]
I think what we have is okay. I'll try to figure out how to add an option to drop some fraction of the energy trajectory in the analysis code, or to allow a user to perform analysis over a subset of the data, which would be useful for convergence testing anyway. I _think_ the pre-processing stuff that Anna added does some of this anyway, but it is hidden in a lot of alchemlyb calls. I think it would be good to have an option to just drop some data if you already know that it's junk.
